File: vec.c
/* See LICENSE file for copyright and license details. */

#include <stdio.h>
#include <stdlib.h>
#include <complex.h>
/* u have to add -lm to the linker */
#include <math.h>

/* get dimension of an array */
#define DIM(X)        (sizeof X / sizeof X[0])

/* define a Vector */
typedef struct {
        int dim;
        double norm;
        double complex *arr;
} Vec;


/* define an option that check for error and return a double*/
typedef struct {
        int error;
        char *msg;
        double result;
} Option_double;


/* define an option that check for error and return a Vector */
typedef struct {
        int error;
        char *msg;
        Vec result;
} Option_vec;


/* get norm of a Vector */
double
get_norm(int dim, double complex* arr)
{
    double norm = 0;
    for (int i = 0; i < dim; i++) {
        norm += pow(cabs(arr[i]), 2);  /* Sum of squared magnitudes */
    }
    return sqrt(norm);
}

/* INIT a Vector */
Vec
new_vec(int dim, double complex* arr)
{
    if (dim <= 0 || arr == NULL) {
        fprintf(stderr, "Invalid vector initialization\n");
        exit(1);
    }

    Vec v = {
        dim,
	get_norm(dim, arr),
        arr  
    };
    return v;
}

void
print_vec(Vec* v, char* name)
{
    printf("Vec %s (\nArray:\n", name);
    for (int i = 0; i < v->dim; i++) {
        printf("\t%f + %fi\n", creal(v->arr[i]), cimag(v->arr[i]));
    }
    printf("Norm:\t%f\n", v->norm);
    printf("Dim:\t%d\n", v->dim);
    puts(")\n");
}

/*
 * Calculates the dot product of two vectors.
 * @param self Pointer to the first vector.
 * @param other Pointer to the second vector.
 * @return Option_double containing the result or an error message.
 */
Option_double
dot_product(Vec* self, Vec* other)
{
    if (self->dim != other->dim) {
        Option_double ret = {1, "Both dimensions must be equal", 0};
        return ret;
    }

    double complex res = 0;
    for (int i = 0; i < self->dim; i++) {
        res += self->arr[i] * conj(other->arr[i]);  /* Dot product with conjugate */
    }

    Option_double ret = {0, "", creal(res)};  /* Return real part of the result */
    return ret;
}

/* define cross product for two 3D Vectors */
Option_vec
cross_product(Vec* self, Vec* other, double* arr)
{
    if (self->dim != 3 || other->dim != 3) {
        Option_vec ret = {1, "Both vectors must be 3-dimensional for cross product", {}};
        return ret;
    }

    arr[0] = self->arr[1] * other->arr[2] - self->arr[2] * other->arr[1];
    arr[1] = self->arr[2] * other->arr[0] - self->arr[0] * other->arr[2];
    arr[2] = self->arr[0] * other->arr[1] - self->arr[1] * other->arr[0];

    Option_vec ret = {0, "", {3, get_norm(3, arr), arr}};
    return ret;
}

/* define addition on tow Vectors */
Option_vec
add_vec(Vec* self, Vec* other, double complex *arr)
{
    if (self->dim != other->dim) {
        Option_vec ret = {1, "Both dimensions must be equal", {}};
        return ret;
    }
    int dim = self->dim;
    for (int i = 0; i < dim; i++) {
        arr[i] = self->arr[i] + other->arr[i];
    }
    Option_vec ret = {0, "", {dim, get_norm(dim, arr), arr}};
    return ret;
}


/* define substitution on tow Vectors */
Option_vec
sub_vec(Vec* self, Vec* other, double *arr)
{
        if (self->dim != other->dim) {
                Option_vec ret = {1, "Both dimensions must be equal", {}};
                return ret;
        }
	int dim = self->dim;
        for (int i=0; i<dim; i++)
                arr[i] = (self->arr[i] - other->arr[i]);
        Option_vec ret = {0, "", {dim, get_norm(dim, arr), arr}};
        return ret;
}

int
isorthogonal(Vec* self, Vec* other)
{
        Option_double dot = dot_product(self, other);
        if (dot.error == 0)
		return dot.result == 0;
	printf("Err: %s\n", dot.msg);
	exit(1);	
}

int
main()
{
        double complex a[] = {0, 1, 0};
        double complex b[] = {0, 0, 1};
	double complex c[DIM(a)];
        Vec v = new_vec(DIM(a), a);
        Vec w = new_vec(DIM(b), b);
        print_vec(&v, "V");
        print_vec(&w, "W");
	puts("Addition of V+W:");
        Option_vec u = add_vec(&v, &w, c);
        if (u.error == 0)
		print_vec(&u.result, "U");
        else {
                printf("Err: %s\n", u.msg);
		exit(1);
	}

        Option_double res = dot_product(&v, &w);
        if (res.error == 0)
                printf("Dot Product of V.W: %f\n", res.result );
        else {
                printf("Err: %s\n", res.msg);
		exit(1);
	}	
	Option_vec cross = cross_product(&v, &w, c);
	if (cross.error == 0)
 	   print_vec(&cross.result, "Cross Product of VxW");
	else {
	    printf("Err: %s\n", cross.msg);
	    exit(1);
	}
	
	printf("Are V & W orthogonal? %d\n", isorthogonal(&v, &w));
	return 0;
}

File: vec.py
from __future__ import annotations
from typing import Any
from math import sqrt


# define a Vector
class Vec:
    def __init__(self: Vec, arr: list[float], name: str = "") -> None:
        self.arr: list[float] = arr
        self.dim: int = len(arr)
        self.norm: float = Vec.get_norm(arr)
        self.name: str = name

    @staticmethod
    def get_norm(arr: list[float]) -> float:
        return sqrt(sum(map(lambda x: x**2, arr)))

    @staticmethod
    def check_vec(other: Any) -> None:
        if not isinstance(other, Vec):
            raise TypeError("Both operands must be of Vec type")

    def __repr__(self: Vec) -> str:
        return f"Vector {self.name}(\nArray: {self.arr}\nDim: {self.dim}\nNorm:{self.norm}\n)"

    def __str__(self: Vec) -> str:
        return f"Vector {self.name}(\nArray: {self.arr}\nDim: {self.dim}\nNorm:{self.norm}\n)"

    def __add__(self: Vec, other: Vec) -> Vec:
        Vec.check_vec(other)
        return Vec([sum(i) for i in zip(self.arr, other.arr)])

    def __sub__(self: Vec, other: Vec) -> Vec:
        Vec.check_vec(other)
        return Vec([i - j for i, j in zip(self.arr, other.arr)])

    def __mul__(self: Vec, other: Vec) -> float:
        if isinstance(other, Vec):
            if self.dim != other.dim:
                raise TypeError(
                    "unsupported operand type(s) for *: unless both vectors have same dimension"
                )
            return sum([i * j for i, j in zip(self.arr, other.arr)])
        elif isinstance(other, int | float):
            return Vec.__rmul__(self, other)
        raise TypeError("Both operands must be of Vec type or scaler * Vec")

    def __rmul__(self: Vec, other: float) -> Vec:
        return Vec([other * i for i in self.arr])
        
    def cross(self: Vec, other: Vec) -> Vec:
        Vec.check_vec(other)
        if self.dim != 3 or other.dim != 3:
            raise ValueError("Cross product is only defined for 3-dimensional vectors")
        cross_product = [
            self.arr[1] * other.arr[2] - self.arr[2] * other.arr[1],
            self.arr[2] * other.arr[0] - self.arr[0] * other.arr[2],
            self.arr[0] * other.arr[1] - self.arr[1] * other.arr[0],
        ]
        return Vec(cross_product, f"{self.name}×{other.name}")    
        
    def isorthogonal(self: Vec, other: Vec) -> bool:
        return self * other == 0


def main() -> None:
    v = Vec([0, 1, 0], "V")
    w = Vec([1, 0, 0], "W")
    u: Vec = v + w
    u.name: str = "U"
    cross_vw = v.cross(w)
    print(v)
    print(w)
    print(u)
    print("Dot product of V.U: ", v * u)
    print("Are V & W orthogonal? ", v.isorthogonal(w))
    print(cross_vw) 


if __name__ == "__main__":
    main()
